File: Solution Union Find.cpp

#include <vector>
#include <unordered_map>
using namespace std;

class UnionFind {
    
    vector<int> parent;
    vector<int> rank;

public:
    UnionFind(int numberOfNodes) {
        parent.resize(numberOfNodes);
        rank.resize(numberOfNodes);
        iota(parent.begin(), parent.end(), 0);
    }

    int findParent(int index) {
        if (parent[index] != index) {
            parent[index] = findParent(parent[index]);
        }
        return parent[index];
    }

    void joinByRank(int first, int second) {
        if (rank[first] > rank[second]) {
            parent[second] = first;
        } else if (rank[first] < rank[second]) {
            parent[first] = second;
        } else {
            parent[second] = first;
            ++rank[first];
        }
    }
};

class Solution {
    
public:
    long long countPairs(int numberOfNodes, vector<vector<int>>& edges) {

        UnionFind unionFind(numberOfNodes);
        joinNodesOfSameGroup(unionFind, edges);

        unordered_map<int, int> parentAndGroupSize;
        fillMap_parentAndGroupSize(unionFind, parentAndGroupSize, numberOfNodes);

        return findNumberOfUnreachablePairsOfNodes(parentAndGroupSize, numberOfNodes);
    }

private:
    long long findNumberOfUnreachablePairsOfNodes(const unordered_map<int, int>& parentAndGroupSize, int numberOfNodes) {
        long long numberOfUnreachablePairsOfNodes = 0;
        int remainingNodesToMakePairs = numberOfNodes;

        for (const auto& groups : parentAndGroupSize) {
            numberOfUnreachablePairsOfNodes += static_cast<long long> (groups.second) * (remainingNodesToMakePairs - groups.second);
            remainingNodesToMakePairs -= groups.second;
        }
        return numberOfUnreachablePairsOfNodes;
    }

    void joinNodesOfSameGroup(UnionFind& unionFind, const vector<vector<int>>& edges) {
        for (const auto& edge : edges) {
            int parentFirst = unionFind.findParent(edge[0]);
            int parentSecond = unionFind.findParent(edge[1]);
            if (parentFirst != parentSecond) {
                unionFind.joinByRank(parentFirst, parentSecond);
            }
        }
    }

    void fillMap_parentAndGroupSize(UnionFind& unionFind, unordered_map<int, int>& parentAndGroupSize, int numberOfNodes) {
        for (int node = 0; node < numberOfNodes; ++node) {
            int parent = unionFind.findParent(node);
            parentAndGroupSize[parent] = parentAndGroupSize[parent] + 1;
        }
    }
};

File: Solution Union Find.js

/**
 * @param {number} numberOfNodes
 * @param {number[][]} edges
 * @return {number}
 */
var countPairs = function (numberOfNodes, edges) {
    const unionFind = new UnionFind(numberOfNodes);
    joinNodesOfSameGroup(unionFind, edges);

    const parentAndGroupSize = new Map();
    fillMap_parentAndGroupSize(unionFind, parentAndGroupSize, numberOfNodes);

    return findNumberOfUnreachablePairsOfNodes(parentAndGroupSize, numberOfNodes);
};

/**
 * @param {Map<number, number>} parentAndGroupSize
 * @param {number} numberOfNodes
 * @return {number}
 */
function findNumberOfUnreachablePairsOfNodes(parentAndGroupSize, numberOfNodes) {
    let numberOfUnreachablePairsOfNodes = 0;
    let remainingNodesToMakePairs = numberOfNodes;

    for (let groups of parentAndGroupSize.values()) {
        numberOfUnreachablePairsOfNodes += (remainingNodesToMakePairs - groups) * groups;
        remainingNodesToMakePairs -= groups;
    }
    return numberOfUnreachablePairsOfNodes;
}

/**
 * @param {class UnionFind} unionFind
 * @param {number[][]} edges
 * @return {void}
 */
function joinNodesOfSameGroup(unionFind, edges) {
    for (let edge of edges) {
        let parentFirst = unionFind.findParent(edge[0]);
        let parentSecond = unionFind.findParent(edge[1]);
        if (parentFirst !== parentSecond) {
            unionFind.joinByRank(parentFirst, parentSecond);
        }
    }
}

/**
 * @param {class UnionFind} unionFind
 * @param {Map<number, number>} parentAndGroupSize
 * @param {number} numberOfNodes
 * @return {void}
 */
function fillMap_parentAndGroupSize(unionFind, parentAndGroupSize, numberOfNodes) {
    for (let node = 0; node < numberOfNodes; ++node) {
        let parent = unionFind.findParent(node);
        if (!parentAndGroupSize.has(parent)) {
            parentAndGroupSize.set(parent, 0);
        }
        parentAndGroupSize.set(parent, parentAndGroupSize.get(parent) + 1);
    }
}

class UnionFind {

    constructor(numberOfNodes) {
        this.parent = Array.from(new Array(numberOfNodes).keys());
        this.rank = new Array(numberOfNodes);
    }

    findParent(index) {
        if (this.parent[index] !== index) {
            this.parent[index] = this.findParent(this.parent[index]);
        }
        return this.parent[index];
    }

    joinByRank(first, second) {
        if (this.rank[first] > this.rank[second]) {
            this.parent[second] = first;
        } else if (this.rank[first] < this.rank[second]) {
            this.parent[first] = second;
        } else {
            this.parent[second] = first;
            ++this.rank[first];
        }
    }
}
